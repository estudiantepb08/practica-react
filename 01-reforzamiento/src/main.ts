import './style.css';
import './bases/01-const-let';
import './bases/02-template-string';
import './bases/03-object-literal.ts';
import './bases/04-arrays.ts';
import './bases/05-funtions.ts';
import './bases/06-obj-destructuring.ts';
import './bases/07-array-destructuring.ts';
import { getHeroeByOwner } from './bases/08-imp-exp.ts'
import { Owner } from './data/heroes.data.ts';
import './bases/09-promises.ts';
import './bases/10-fetch-api.ts';
import './bases/11-async-await.ts';

document.querySelector<HTMLDivElement>('#app')!.innerHTML = `
  <div>
    <h1>HOLA MUNDO</h1>
  </div>
`
console.log(getHeroeByOwner(Owner.Marvel));
console.log(getHeroeByOwner(Owner.DC));

// Importante: Si se importa algo que no se usa, da error. Por eso se importan los otros archivos

// Ejemplos de desestructuraci√≥n


