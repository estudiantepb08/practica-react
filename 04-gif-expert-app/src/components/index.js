/**Es un archivo de barril para utilizar todos los componentes y hacer una solo importacion como lo hace react */

export * from './AddCategories';
export * from './GifGrid';
export * from './GifItem';