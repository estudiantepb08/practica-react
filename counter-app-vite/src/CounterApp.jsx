import React from 'react';
import { useState } from 'react';

export const CounterApp = ({value})=>{
    /**
     * Creamos nuestro primer hook que es el estado el hook renderiza o dibuja el nuevo estado del componente
     * utiliza los argumentos el valor que va a dibujar y donde lo voy a setear con set()
     * siempre deben ser una constante
     * const [counter, setCounter] = useState(0);
     */
    const [counter, setCounter] = useState(value);

    const handleApp = () =>{setCounter(counter + 1)};
    const handleSubtract = () =>{setCounter(counter - 1)};
    const handleReset = () =>{setCounter(value)};
    /*
    se puede enviar solo la referencia a la funcion sin enviarle la cordenada completa
    tradicional
    <button onClick={(event) => handleApp(event)}> + 1</button>
    solo referencia en la funcion se pasan los argumentos de la misma forma
    <button onClick={handleApp}> + 1</button>
    */
   /**
    * Para realizar las pruebas unitarias necesitamos installar en VITE la libreria de JEST
    * Y en el package.json agregar la propiedad "test":"jest" "test": "jest --watchAll"
    * para ejecutar las pruebas es con el comando yarn test o npm run test 
    * otra manera de identificar un elemento HTML es con aria-label
    */

    return(
        <>
        <h1>Evento clic</h1>
        <h2>{ counter }</h2>
        <button onClick={ handleApp }> + 1</button>
        <button onClick={ handleSubtract }> - 1</button>
        <button aria-label='btn-reset' onClick={ handleReset }> Reset</button>
        </>
    );
}
