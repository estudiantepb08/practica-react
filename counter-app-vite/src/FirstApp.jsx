import React from 'react';

const newMessages = 'ALEX'; 
/*
esta constantes deben inr fuera de la funcion para que no se ha renderizado
No he permitodo los objetos json, solo es permitido serializandolo con la clase JSON.stringify
*/

const getResultado = (a,b) =>{
    return a + b;
};

export const FirstApp = () =>{
    return (
        <>
    <h2>First App</h2>
    <p>hola: { newMessages }</p>
    <p>Valor Numero: {getResultado(2,8)}</p>
    </>
    );
};