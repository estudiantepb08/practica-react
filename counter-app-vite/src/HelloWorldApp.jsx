import React from 'react';

export function App(){
    return (<h1>Hellow Work</h1>)
}

export default App;
