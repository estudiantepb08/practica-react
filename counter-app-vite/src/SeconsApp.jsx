import React from 'react';
import PropTypes from 'prop-types';

export const SeconsApp = ({title, subtitle}) =>{
    /**
     * Las props se pueden desestructurar colocandolas entre llaves {} o solas
     * es decir el nombre del objecto dentro del argumento del componente
     */
    return (<>
    <h1 data-testid="test-title"> { title } </h1>
    <p>Mis primeras props titulo: Trabajando Con Las Props O Propiedades De La Aplicaci√≥n:</p>
    <p>Mis primeras props subtitulo: { subtitle +1}</p>
    </>);
};

/**
 * Aqui se deben definir las propiedades que va a utilizar el componente
 * se define el tipo de variable, si es obligatoria con la propiedad .isRequired
 */
SeconsApp.propTypes = {
    title: PropTypes.string.isRequired,
    subtitle: PropTypes.number.isRequired
};
/**
 * Tambien podemos definir propiedades por defecto con el props * 
 */

SeconsApp.defaultProps ={
    title: 'Ejemplo No hay tiulo',
    /*subtitle: 23*/
}