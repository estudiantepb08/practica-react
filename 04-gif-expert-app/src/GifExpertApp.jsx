import React from 'react';

import { useState } from "react";
import { AddCategories, GifGrid } from './components';

export const GifExpertApp = ()=>{

    const [categories, setCategories] = useState(['One Punch','Saludo']); 
    /** El valor inicial que pase en el useState lo obtengo con el segundo argumento es decir
     * categories seria mi getCategories, y setCategories me recibe el valor
     */

    const onAddCategory = (newCategories)=>{
        /**Validamos que la categoria no este repetidad en el arreglo */
        if(categories.includes(newCategories)) return;
        //categories.push('Que tal');
        setCategories([newCategories, ...categories]); // hacer copia de categoria y agregar el nuevo valor 'Forma express'
        //setCategories(cat => [...cat, 'valorant']); // con callback
//<button onClick={onAddCategory}>Agregar</button>
    };

    return(<>
    {/** Titulo */}
    <h1>GitExpertApp</h1>
    {/** Input */}
    <AddCategories onNewCategories={onAddCategory}></AddCategories>
    {/** Listado de Gif */}
    
    <ol>{
       /* Primera forma de poder listar la lista de categorias
        categories.map(category =>{
            return<li key={category}>{category}</li>
        })
        */
       /** Segunda forma de listar o recorrer la lista desestructurando con un componente */

       categories.map(category =>(
        <GifGrid key={category} category={category}></GifGrid> //Componente GifGrid categori para desacoplar el codigo
       ))
        }
        
        {/*<li>123</li>
        <li>XYZ</li>*/}
    </ol>
    {/** Gif Item */}
    </>);
}