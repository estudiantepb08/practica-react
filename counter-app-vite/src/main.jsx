import ReactDOM from 'react-dom'
import React from 'react'
import { createRoot } from 'react-dom/client';
import App from './HelloWorldApp';

import moduleName from 'module'
import { FirstApp } from './FirstApp';
import './styles.css'; // importar hoja de estilos de manera global para todos los componentes
import { SeconsApp } from './SeconsApp';
import { CounterApp } from './CounterApp';

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        {/*<App></App>
        <FirstApp></FirstApp>
        <SeconsApp title='PROPS123' subtitle={12}></SeconsApp>
        */
        }
        
        <CounterApp value={2}></CounterApp>
        
    </React.StrictMode>
);
