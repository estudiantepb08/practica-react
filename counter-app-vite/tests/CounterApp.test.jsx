import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { CounterApp } from "../src/CounterApp";

describe('Pruebas del componente <CounterApp/>', () =>{

    const valor = 2;

    test('Pruebas debe hacer match con el snapshot', ()=>{
        const {container} = render(<CounterApp value={valor}/>);
        expect(container).toMatchSnapshot();  

    });

    test('Pruebas debe mostrar el valor inicial de 100', ()=>{
        render(<CounterApp value={100}></CounterApp>);
        expect(screen.getByText(100)).toBeTruthy();
        /** Otra opciÃ³n */
        expect(screen.getByRole('heading',{level:2}).innerHTML).toContain('100');
    });

    test('Prueba debe mostrar el evento clic',() =>{
        render(<CounterApp value={valor}></CounterApp>);
        fireEvent.click(screen.getByText('+ 1'));
        screen.debug(); //me permite ver el estado del DOM
        expect(screen.getByText(3)).toBeTruthy();
    });

    test('Prueba debe mostrar el evento clic -1', ()=>{
        render(<CounterApp value={valor}></CounterApp>);
        fireEvent.click(screen.getByText('- 1'));
        screen.debug();
        expect(screen.getByText(1)).toBeTruthy();
    });

    test('Prueba debe funcionar el boton Reset', ()=>{
        render(<CounterApp value={valor}></CounterApp>);
        fireEvent.click(screen.getByText('+ 1'));
        fireEvent.click(screen.getByText('+ 1'));
        fireEvent.click(screen.getByText('+ 1'));
       // fireEvent.click(screen.getByText('Reset'));
       /** Podemos tomar el boton con el atributo aria-label */
        fireEvent.click(screen.getByRole('button', {name:'btn-reset'}));
        expect(screen.getByText(2)).toBeTruthy();
    });
});