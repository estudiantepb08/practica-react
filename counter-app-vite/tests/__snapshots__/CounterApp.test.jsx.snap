// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <CounterApp/> Pruebas debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    Evento clic
  </h1>
  <h2>
    2
  </h2>
  <button>
     + 1
  </button>
  <button>
     - 1
  </button>
  <button
    aria-label="btn-reset"
  >
     Reset
  </button>
</div>
`;
